src/events/masp.upsert.service.ts

```diff
@@ -0,0 +1,190 @@
+    private readonly usersService: UsersService,
+  ) {}
+
+  async bulkUpsert(
+    operations: UpsertMaspTransactionsOperationDto[],
+  ): Promise<void> {
+    for (const operation of operations) {
+      // We only want to handle masp transactions that deal with the main chain
+      // (not forks). This will only be connected and disconnected events
+      const shouldUpsertMaspTransactions =
+        operation.type === BlockOperation.CONNECTED ||
+        operation.type === BlockOperation.DISCONNECTED;
+
+      if (shouldUpsertMaspTransactions) {
+        await this.upsert(operation);
+      }
+    }
+  }
+
+  async upsert(
+    operation: UpsertMaspTransactionsOperationDto,
+  ): Promise<MaspTransaction[]> {
+    const networkVersion = this.config.get<number>('NETWORK_VERSION');
+    const blockHash = standardizeHash(operation.block.hash);
+
+    assert(
+      operation.type === BlockOperation.CONNECTED ||
+        operation.type === BlockOperation.DISCONNECTED,
+      'FORK operations not supported',
+    );
+
+    const userGraffitis = operation.transactions.map(
+      (transaction) => transaction.assetName,
+    );
+    const users = await this.usersService.findManyAndMapByGraffiti(
+      userGraffitis,
+    );
+
+    const previousBlockHash = standardizeHash(
+      operation.block.previousBlockHash,
+    );
+
+    const maspTransactions = await this.prisma.$transaction(async (prisma) => {
+      const head = await this.maspTransactionHeadService.head();
+      let maspTransactions = new Array<MaspTransaction>();
+      if (operation.type === BlockOperation.CONNECTED) {
+        if (head && head.block_hash !== previousBlockHash) {
+          throw new Error(
+            `Cannot connect block ${blockHash} to ${String(
+              head.block_hash,
+            )}, expecting ${previousBlockHash}`,
+          );
+        }
+        // Create masp transaction params
+        const maspTransactionParams = new Array<{
+          transaction_hash: string;
+          block_hash: string;
+          block_sequence: number;
+          network_version: number;
+          type: EventType;
+          asset_name: string;
+          main: boolean;
+        }>();
+        for (const transaction of operation.transactions) {
+          // Masp assetName should match user grafitti
+          if (!users.has(transaction.assetName)) {
+            continue;
+          }
+
+          maspTransactionParams.push({
+            asset_name: transaction.assetName,
+            type: transaction.type,
+            transaction_hash: standardizeHash(transaction.hash),
+            block_hash: blockHash,
+            block_sequence: operation.block.sequence,
+            main: true,
+            network_version: networkVersion,
+          });
+        }
+
+        // Now create new not existing masp transactions
+        await prisma.maspTransaction.createMany({
+          data: maspTransactionParams,
+          skipDuplicates: true,
+        });
```

Comment:
This will crash during a re-org because of the unique on transactionHash.

Reply:
It will silently fail and not update the masp transactions, which may be worse than crash 😅